pkg/bid/bid.go

```diff
@@ -0,0 +1,149 @@
+package bid
+
+import (
+	"fmt"
+	"strings"
+
+	v2 "github.com/conductorone/baton-sdk/pb/c1/connector/v2"
+)
+
+const (
+	BidPrefix            = "bid:"
+	ResourceBidPrefix    = BidPrefix + "r:"
+	EntitlementBidPrefix = BidPrefix + "e:"
+	GrantBidPrefix       = BidPrefix + "g:"
+)
+
+func ToBid(b interface{}) string {
+	switch bType := b.(type) {
+	case *v2.Resource:
+		return ResourceToBid(bType)
+	case *v2.Entitlement:
+		return EntitlementToBid(bType)
+	case *v2.Grant:
+		return GrantToBid(bType)
+	}
+	return "☠️" // TODO: panic? this should never happen
+}
+
+func ParseBid(bidStr string) (interface{}, error) {
+	if len(bidStr) < 6 {
+		return nil, fmt.Errorf("baton id string too short: %s", bidStr)
+	}
+	bidPrefix := bidStr[:4]
+	if bidPrefix != BidPrefix {
+		return nil, fmt.Errorf("invalid baton id prefix: %s", bidPrefix)
+	}
+
+	bidType := bidStr[:6]
+	switch bidType {
+	case ResourceBidPrefix:
+		return ParseResourceBid(bidStr)
+	case EntitlementBidPrefix:
+		return ParseEntitlementBid(bidStr)
+	case GrantBidPrefix:
+		return ParseGrantBid(bidStr)
+	}
+	return nil, fmt.Errorf("invalid baton id type: %s", bidType)
+}
+
+func parseResourcePart(resourcePart string) (*v2.Resource, error) {
+	resourceId := &v2.ResourceId{}
+	parentResourceId := &v2.ResourceId{}
+	parts := strings.Split(resourcePart, "/")
```

Comment:
The format here is pretty simple so this is probably overkill/code golf-y, but:

I'm generally a fan of the "lexical scanning" pattern for parsing things like this. Keep some state and scan character-by-character.

It lets you return really detailed errors, really simplifies things like handling escapes, and makes it easy to implement things like arbitrary depth here in this thing. (Are we *sure* we want to limit to exactly 1 depth of hierarchy?)